test/support/utils_sycl.h

```diff
@@ -139,28 +139,88 @@ make_new_policy(_Policy&& __policy)
 // create the queue with custom asynchronous exceptions handler
 static auto my_queue = sycl::queue(default_selector, async_handler);
 
-// Invoke op(policy,rest...) for each possible policy.
+// check fp16/fp64 support by a device
+template<typename T>
+bool has_type_support(const sycl::device&) { return true; }
+
+template<>
+inline bool has_type_support<double>(const sycl::device& device)
+{
+    return device.has(sycl::aspect::fp64);
+}
+
+template<>
+inline bool has_type_support<sycl::half>(const sycl::device& device)
+{
+    return device.has(sycl::aspect::fp16);
+}
+
+template <typename, typename = void>
+struct has_value_type : ::std::false_type {};
+
+template <typename T>
+struct has_value_type<T, ::std::void_t<typename std::iterator_traits<T>::value_type>> : ::std::true_type {};
+
+template<typename T>
+std::enable_if_t<!has_value_type<T>::value, bool> has_iter_type_support(const sycl::device& device)
```

Comment:
Finally, we have to do runtime check, right () see line 155?
So, it seems we don't need  any SFINAE compile time dispatching.
An approach based on overloading is more preferable here, IMHO.
(https://godbolt.org/z/dbaoe74se)